File: golang/06-nested-if-hell/loan_approver_test.go
package kata

import "testing"

func TestCanDeliverAllowsNormalPackage(t *testing.T) {
	la := NewLoanApprover()
	pkg := &Package{Weight: 10, HasWeight: true, Hazardous: false, Weekend: false}
	result := la.can_deliver(pkg)
	if !result.Allowed {
		t.Errorf("expected allowed true, got false")
	}
	if result.Warning != "" {
		t.Errorf("expected no warning, got %q", result.Warning)
	}
}

func TestCanDeliverRejectsOverweight(t *testing.T) {
	la := NewLoanApprover()
	pkg := &Package{Weight: 60, HasWeight: true, Hazardous: false, Weekend: false}
	result := la.can_deliver(pkg)
	if result.Allowed {
		t.Errorf("expected allowed false, got true")
	}
	if result.Warning != "Weight exceeded" {
		t.Errorf("expected warning 'Weight exceeded', got %q", result.Warning)
	}
}

func TestCanDeliverRejectsHazardous(t *testing.T) {
	la := NewLoanApprover()
	pkg := &Package{Weight: 10, HasWeight: true, Hazardous: true, Weekend: false}
	result := la.can_deliver(pkg)
	if result.Allowed {
		t.Errorf("expected allowed false, got true")
	}
	if result.Warning != "Hazardous material" {
		t.Errorf("expected warning 'Hazardous material', got %q", result.Warning)
	}
}

func TestCanDeliverRejectsWeekend(t *testing.T) {
	la := NewLoanApprover()
	pkg := &Package{Weight: 10, HasWeight: true, Hazardous: false, Weekend: true}
	result := la.can_deliver(pkg)
	if result.Allowed {
		t.Errorf("expected allowed false, got true")
	}
	if result.Warning != "No weekend delivery" {
		t.Errorf("expected warning 'No weekend delivery', got %q", result.Warning)
	}
}

func TestCanDeliverRejectsExtremeTemperature(t *testing.T) {
	la := NewLoanApprover()
	temp50 := 50
	pkg := &Package{Weight: 10, HasWeight: true, Hazardous: false, Weekend: false, TemperatureRequired: &temp50}
	result := la.can_deliver(pkg)
	if result.Allowed {
		t.Errorf("expected allowed false, got true")
	}
	if result.Warning != "Temperature out of range" {
		t.Errorf("expected warning 'Temperature out of range', got %q", result.Warning)
	}
}

func TestCanDeliverAllowsValidTemperature(t *testing.T) {
	la := NewLoanApprover()
	temp20 := 20
	pkg := &Package{Weight: 10, HasWeight: true, Hazardous: false, Weekend: false, TemperatureRequired: &temp20}
	result := la.can_deliver(pkg)
	if !result.Allowed {
		t.Errorf("expected allowed true, got false")
	}
}

func TestCanDeliverAllowsRemoteSmallPackage(t *testing.T) {
	la := NewLoanApprover()
	pkg := &Package{Weight: 15, HasWeight: true, Hazardous: false, Weekend: false, RemoteArea: true}
	result := la.can_deliver(pkg)
	if !result.Allowed {
		t.Errorf("expected allowed true, got false")
	}
	if result.Warning != "Remote surcharge applies" {
		t.Errorf("expected warning 'Remote surcharge applies', got %q", result.Warning)
	}
}

func TestCanDeliverRejectsRemoteHeavyPackage(t *testing.T) {
	la := NewLoanApprover()
	pkg := &Package{Weight: 25, HasWeight: true, Hazardous: false, Weekend: false, RemoteArea: true}
	result := la.can_deliver(pkg)
	if result.Allowed {
		t.Errorf("expected allowed false, got true")
	}
	if result.Warning != "Too heavy for remote" {
		t.Errorf("expected warning 'Too heavy for remote', got %q", result.Warning)
	}
}

func TestCanDeliverRejectsNullPackage(t *testing.T) {
	la := NewLoanApprover()
	result := la.can_deliver(nil)
	if result.Allowed {
		t.Errorf("expected allowed false, got true")
	}
	if result.Warning != "No package" {
		t.Errorf("expected warning 'No package', got %q", result.Warning)
	}
}

func TestCanDeliverRejectsMissingWeight(t *testing.T) {
	la := NewLoanApprover()
	pkg := &Package{Hazardous: false}
	result := la.can_deliver(pkg)
	if result.Allowed {
		t.Errorf("expected allowed false, got true")
	}
	if result.Warning != "No weight specified" {
		t.Errorf("expected warning 'No weight specified', got %q", result.Warning)
	}
}

File: csharp/06-nested-if-hell/test/loan_approver_test.cs
using Xunit;

public class LoanApproverTest
{
    [Fact]
    public void AllowsNormalPackage()
    {
        var checker = new LoanApprover();
        var result = checker.can_deliver(new Package { Weight = 10, Hazardous = false, Weekend = false });
        Assert.True(result.Allowed);
        Assert.Null(result.Warning);
    }

    [Fact]
    public void RejectsOverweight()
    {
        var checker = new LoanApprover();
        var result = checker.can_deliver(new Package { Weight = 60, Hazardous = false, Weekend = false });
        Assert.False(result.Allowed);
        Assert.Equal("Weight exceeded", result.Warning);
    }

    [Fact]
    public void RejectsHazardous()
    {
        var checker = new LoanApprover();
        var result = checker.can_deliver(new Package { Weight = 10, Hazardous = true, Weekend = false });
        Assert.False(result.Allowed);
        Assert.Equal("Hazardous material", result.Warning);
    }

    [Fact]
    public void RejectsWeekend()
    {
        var checker = new LoanApprover();
        var result = checker.can_deliver(new Package { Weight = 10, Hazardous = false, Weekend = true });
        Assert.False(result.Allowed);
        Assert.Equal("No weekend delivery", result.Warning);
    }

    [Fact]
    public void RejectsExtremeTemperature()
    {
        var checker = new LoanApprover();
        var result = checker.can_deliver(new Package { Weight = 10, Hazardous = false, Weekend = false, TemperatureRequired = 50 });
        Assert.False(result.Allowed);
        Assert.Equal("Temperature out of range", result.Warning);
    }

    [Fact]
    public void AllowsValidTemperature()
    {
        var checker = new LoanApprover();
        var result = checker.can_deliver(new Package { Weight = 10, Hazardous = false, Weekend = false, TemperatureRequired = 20 });
        Assert.True(result.Allowed);
    }

    [Fact]
    public void AllowsRemoteSmallPackage()
    {
        var checker = new LoanApprover();
        var result = checker.can_deliver(new Package { Weight = 15, Hazardous = false, Weekend = false, RemoteArea = true });
        Assert.True(result.Allowed);
        Assert.Equal("Remote surcharge applies", result.Warning);
    }

    [Fact]
    public void RejectsRemoteHeavyPackage()
    {
        var checker = new LoanApprover();
        var result = checker.can_deliver(new Package { Weight = 25, Hazardous = false, Weekend = false, RemoteArea = true });
        Assert.False(result.Allowed);
        Assert.Equal("Too heavy for remote", result.Warning);
    }

    [Fact]
    public void RejectsNullPackage()
    {
        var checker = new LoanApprover();
        var result = checker.can_deliver(null);
        Assert.False(result.Allowed);
        Assert.Equal("No package", result.Warning);
    }

    [Fact]
    public void RejectsMissingWeight()
    {
        var checker = new LoanApprover();
        var result = checker.can_deliver(new Package { Hazardous = false, Weekend = false });
        Assert.False(result.Allowed);
        Assert.Equal("No weight specified", result.Warning);
    }
}
